### PR TITLE
using the project-repository map while being in write lock

### LINKED ISSUE #2445: Repositories are doubled on the main page in the midst of parallel reindex
I'm using the setup as described in #2382 with PR #2423 applied and I run the given setup 100 times in a row. It happened to mi twice already that at some point, the repositories are wrongly assigned which leads to multiple repositories per single project, as in the screenshot: 

screen shot 2018-10-25 at 13 06 20

The configuration, as retrieved from the webapp is
```xml
<?xml version="1.0" encoding="UTF-8"?>
<java version="1.8.0_144" class="java.beans.XMLDecoder">
 <object class="org.opengrok.indexer.configuration.Configuration" id="Configuration0">
  <void property="cmds">
   <object class="java.util.Collections" method="unmodifiableMap">
    <object class="java.util.HashMap">
     <void method="put">
      <string>org.opengrok.indexer.history.SubversionRepository</string>
      <string>/usr/bin/svn</string>
     </void>
     <void method="put">
      <string>org.opengrok.indexer.history.AccuRevRepository</string>
      <string>accurev</string>
     </void>
     <void method="put">
      <string>org.opengrok.indexer.history.ClearCaseRepository</string>
      <string>cleartool</string>
     </void>
     <void method="put">
      <string>org.opengrok.indexer.history.PerforceRepository</string>
      <string>p4</string>
     </void>
     <void method="put">
      <string>org.opengrok.indexer.history.GitRepository</string>
      <string>/usr/bin/git</string>
     </void>
    </object>
   </object>
  </void>
  <void property="ctags">
   <string>/usr/local/bin/ctags</string>
  </void>
  <void property="dataRoot">
   <string>/private/tmp/opengrok/data</string>
  </void>
  <void property="groups">
   <void method="add">
    <object class="org.opengrok.indexer.configuration.Group" id="Group0">
     <void property="flag">
      <int>1</int>
     </void>
     <void property="name">
      <string>tulinkry</string>
     </void>
     <void property="pattern">
      <string>tulinkry.*</string>
     </void>
     <void property="repositories">
      <void method="add">
       <object class="org.opengrok.indexer.configuration.Project" id="Project0">
        <void property="groups">
         <void method="add">
          <object idref="Group0"/>
         </void>
        </void>
        <void property="historyEnabled">
         <boolean>true</boolean>
        </void>
        <void property="indexed">
         <boolean>true</boolean>
        </void>
        <void property="name">
         <string>tulinkry.git.1</string>
        </void>
        <void property="path">
         <string>/tulinkry.git.1</string>
        </void>
       </object>
      </void>
      <void method="add">
       <object class="org.opengrok.indexer.configuration.Project" id="Project1">
        <void property="groups">
         <void method="add">
          <object idref="Group0"/>
         </void>
        </void>
        <void property="historyEnabled">
         <boolean>true</boolean>
        </void>
        <void property="indexed">
         <boolean>true</boolean>
        </void>
        <void property="name">
         <string>tulinkry.git.10</string>
        </void>
        <void property="path">
         <string>/tulinkry.git.10</string>
        </void>
       </object>
      </void>
      <void method="add">
       <object class="org.opengrok.indexer.configuration.Project" id="Project2">
        <void property="groups">
         <void method="add">
          <object idref="Group0"/>
         </void>
        </void>
        <void property="historyEnabled">
         <boolean>true</boolean>
        </void>
        <void property="indexed">
         <boolean>true</boolean>
        </void>
        <void property="name">
         <string>tulinkry.git.2</string>
        </void>
        <void property="path">
         <string>/tulinkry.git.2</string>
        </void>
       </object>
      </void>
      <void method="add">
       <object class="org.opengrok.indexer.configuration.Project" id="Project3">
        <void property="groups">
         <void method="add">
          <object idref="Group0"/>
         </void>
        </void>
        <void property="historyEnabled">
         <boolean>true</boolean>
        </void>
        <void property="indexed">
         <boolean>true</boolean>
        </void>
        <void property="name">
         <string>tulinkry.git.3</string>
        </void>
        <void property="path">
         <string>/tulinkry.git.3</string>
        </void>
       </object>
      </void>
      <void method="add">
       <object class="org.opengrok.indexer.configuration.Project" id="Project4">
        <void property="groups">
         <void method="add">
          <object idref="Group0"/>
         </void>
        </void>
        <void property="historyEnabled">
         <boolean>true</boolean>
        </void>
        <void property="indexed">
         <boolean>true</boolean>
        </void>
        <void property="name">
         <string>tulinkry.git.4</string>
        </void>
        <void property="path">
         <string>/tulinkry.git.4</string>
        </void>
       </object>
      </void>
      <void method="add">
       <object class="org.opengrok.indexer.configuration.Project" id="Project5">
        <void property="groups">
         <void method="add">
          <object idref="Group0"/>
         </void>
        </void>
        <void property="historyEnabled">
         <boolean>true</boolean>
        </void>
        <void property="indexed">
         <boolean>true</boolean>
        </void>
        <void property="name">
         <string>tulinkry.git.5</string>
        </void>
        <void property="path">
         <string>/tulinkry.git.5</string>
        </void>
       </object>
      </void>
      <void method="add">
       <object class="org.opengrok.indexer.configuration.Project" id="Project6">
        <void property="groups">
         <void method="add">
          <object idref="Group0"/>
         </void>
        </void>
        <void property="historyEnabled">
         <boolean>true</boolean>
        </void>
        <void property="indexed">
         <boolean>true</boolean>
        </void>
        <void property="name">
         <string>tulinkry.git.6</string>
        </void>
        <void property="path">
         <string>/tulinkry.git.6</string>
        </void>
       </object>
      </void>
      <void method="add">
       <object class="org.opengrok.indexer.configuration.Project" id="Project7">
        <void property="groups">
         <void method="add">
          <object idref="Group0"/>
         </void>
        </void>
        <void property="historyEnabled">
         <boolean>true</boolean>
        </void>
        <void property="indexed">
         <boolean>true</boolean>
        </void>
        <void property="name">
         <string>tulinkry.git.7</string>
        </void>
        <void property="path">
         <string>/tulinkry.git.7</string>
        </void>
       </object>
      </void>
      <void method="add">
       <object class="org.opengrok.indexer.configuration.Project" id="Project8">
        <void property="groups">
         <void method="add">
          <object idref="Group0"/>
         </void>
        </void>
        <void property="historyEnabled">
         <boolean>true</boolean>
        </void>
        <void property="indexed">
         <boolean>true</boolean>
        </void>
        <void property="name">
         <string>tulinkry.git.8</string>
        </void>
        <void property="path">
         <string>/tulinkry.git.8</string>
        </void>
       </object>
      </void>
      <void method="add">
       <object class="org.opengrok.indexer.configuration.Project" id="Project9">
        <void property="groups">
         <void method="add">
          <object idref="Group0"/>
         </void>
        </void>
        <void property="historyEnabled">
         <boolean>true</boolean>
        </void>
        <void property="indexed">
         <boolean>true</boolean>
        </void>
        <void property="name">
         <string>tulinkry.git.9</string>
        </void>
        <void property="path">
         <string>/tulinkry.git.9</string>
        </void>
       </object>
      </void>
     </void>
    </object>
   </void>
  </void>
  <void property="projects">
   <void method="put">
    <string>tulinkry.git.5</string>
    <object idref="Project5"/>
   </void>
   <void method="put">
    <string>tulinkry.git.4</string>
    <object idref="Project4"/>
   </void>
   <void method="put">
    <string>tulinkry.git.7</string>
    <object idref="Project7"/>
   </void>
   <void method="put">
    <string>tulinkry.git.6</string>
    <object idref="Project6"/>
   </void>
   <void method="put">
    <string>tulinkry.git.1</string>
    <object idref="Project0"/>
   </void>
   <void method="put">
    <string>tulinkry.git.3</string>
    <object idref="Project3"/>
   </void>
   <void method="put">
    <string>tulinkry.git.2</string>
    <object idref="Project2"/>
   </void>
   <void method="put">
    <string>tulinkry.git.9</string>
    <object idref="Project9"/>
   </void>
   <void method="put">
    <string>tulinkry.git.10</string>
    <object idref="Project1"/>
   </void>
   <void method="put">
    <string>tulinkry.git.8</string>
    <object idref="Project8"/>
   </void>
  </void>
  <void property="projectsEnabled">
   <boolean>true</boolean>
  </void>
  <void property="repositories">
   <void method="add">
    <object class="org.opengrok.indexer.history.RepositoryInfo">
     <void property="branch">
      <string>master</string>
     </void>
     <void property="currentVersion">
      <string>2018-10-22 19:49 +0200 675395c Kryštof Tulinger fixing composer.json</string>
     </void>
     <void property="datePatterns">
      <array class="java.lang.String" length="1">
       <void index="0">
        <string>yyyy-MM-dd&apos;T&apos;HH:mm:ssXXX</string>
       </void>
      </array>
     </void>
     <void property="directoryNameRelative">
      <string>/tulinkry.git.1</string>
     </void>
     <void property="parent">
      <string>https://github.com/tulinkry/git</string>
     </void>
     <void property="type">
      <string>git</string>
     </void>
    </object>
   </void>
   <void method="add">
    <object class="org.opengrok.indexer.history.RepositoryInfo">
     <void property="branch">
      <string>master</string>
     </void>
     <void property="currentVersion">
      <string>2018-10-22 19:49 +0200 675395c Kryštof Tulinger fixing composer.json</string>
     </void>
     <void property="datePatterns">
      <array class="java.lang.String" length="1">
       <void index="0">
        <string>yyyy-MM-dd&apos;T&apos;HH:mm:ssXXX</string>
       </void>
      </array>
     </void>
     <void property="directoryNameRelative">
      <string>/tulinkry.git.10</string>
     </void>
     <void property="parent">
      <string>https://github.com/tulinkry/git</string>
     </void>
     <void property="type">
      <string>git</string>
     </void>
    </object>
   </void>
   <void method="add">
    <object class="org.opengrok.indexer.history.RepositoryInfo">
     <void property="branch">
      <string>master</string>
     </void>
     <void property="currentVersion">
      <string>2018-10-22 19:49 +0200 675395c Kryštof Tulinger fixing composer.json</string>
     </void>
     <void property="datePatterns">
      <array class="java.lang.String" length="1">
       <void index="0">
        <string>yyyy-MM-dd&apos;T&apos;HH:mm:ssXXX</string>
       </void>
      </array>
     </void>
     <void property="directoryNameRelative">
      <string>/tulinkry.git.7</string>
     </void>
     <void property="parent">
      <string>https://github.com/tulinkry/git</string>
     </void>
     <void property="type">
      <string>git</string>
     </void>
    </object>
   </void>
   <void method="add">
    <object class="org.opengrok.indexer.history.RepositoryInfo">
     <void property="branch">
      <string>master</string>
     </void>
     <void property="currentVersion">
      <string>2018-10-22 19:49 +0200 675395c Kryštof Tulinger fixing composer.json</string>
     </void>
     <void property="datePatterns">
      <array class="java.lang.String" length="1">
       <void index="0">
        <string>yyyy-MM-dd&apos;T&apos;HH:mm:ssXXX</string>
       </void>
      </array>
     </void>
     <void property="directoryNameRelative">
      <string>/tulinkry.git.6</string>
     </void>
     <void property="parent">
      <string>https://github.com/tulinkry/git</string>
     </void>
     <void property="type">
      <string>git</string>
     </void>
    </object>
   </void>
   <void method="add">
    <object class="org.opengrok.indexer.history.RepositoryInfo">
     <void property="branch">
      <string>master</string>
     </void>
     <void property="currentVersion">
      <string>2018-10-22 19:49 +0200 675395c Kryštof Tulinger fixing composer.json</string>
     </void>
     <void property="datePatterns">
      <array class="java.lang.String" length="1">
       <void index="0">
        <string>yyyy-MM-dd&apos;T&apos;HH:mm:ssXXX</string>
       </void>
      </array>
     </void>
     <void property="directoryNameRelative">
      <string>/tulinkry.git.9</string>
     </void>
     <void property="parent">
      <string>https://github.com/tulinkry/git</string>
     </void>
     <void property="type">
      <string>git</string>
     </void>
    </object>
   </void>
   <void method="add">
    <object class="org.opengrok.indexer.history.RepositoryInfo">
     <void property="branch">
      <string>master</string>
     </void>
     <void property="currentVersion">
      <string>2018-10-22 19:49 +0200 675395c Kryštof Tulinger fixing composer.json</string>
     </void>
     <void property="datePatterns">
      <array class="java.lang.String" length="1">
       <void index="0">
        <string>yyyy-MM-dd&apos;T&apos;HH:mm:ssXXX</string>
       </void>
      </array>
     </void>
     <void property="directoryNameRelative">
      <string>/tulinkry.git.8</string>
     </void>
     <void property="parent">
      <string>https://github.com/tulinkry/git</string>
     </void>
     <void property="type">
      <string>git</string>
     </void>
    </object>
   </void>
   <void method="add">
    <object class="org.opengrok.indexer.history.RepositoryInfo">
     <void property="branch">
      <string>master</string>
     </void>
     <void property="currentVersion">
      <string>2018-10-22 19:49 +0200 675395c Kryštof Tulinger fixing composer.json</string>
     </void>
     <void property="datePatterns">
      <array class="java.lang.String" length="1">
       <void index="0">
        <string>yyyy-MM-dd&apos;T&apos;HH:mm:ssXXX</string>
       </void>
      </array>
     </void>
     <void property="directoryNameRelative">
      <string>/tulinkry.git.3</string>
     </void>
     <void property="parent">
      <string>https://github.com/tulinkry/git</string>
     </void>
     <void property="type">
      <string>git</string>
     </void>
    </object>
   </void>
   <void method="add">
    <object class="org.opengrok.indexer.history.RepositoryInfo">
     <void property="branch">
      <string>master</string>
     </void>
     <void property="currentVersion">
      <string>2018-10-22 19:49 +0200 675395c Kryštof Tulinger fixing composer.json</string>
     </void>
     <void property="datePatterns">
      <array class="java.lang.String" length="1">
       <void index="0">
        <string>yyyy-MM-dd&apos;T&apos;HH:mm:ssXXX</string>
       </void>
      </array>
     </void>
     <void property="directoryNameRelative">
      <string>/tulinkry.git.2</string>
     </void>
     <void property="parent">
      <string>https://github.com/tulinkry/git</string>
     </void>
     <void property="type">
      <string>git</string>
     </void>
    </object>
   </void>
   <void method="add">
    <object class="org.opengrok.indexer.history.RepositoryInfo">
     <void property="branch">
      <string>master</string>
     </void>
     <void property="currentVersion">
      <string>2018-10-22 19:49 +0200 675395c Kryštof Tulinger fixing composer.json</string>
     </void>
     <void property="datePatterns">
      <array class="java.lang.String" length="1">
       <void index="0">
        <string>yyyy-MM-dd&apos;T&apos;HH:mm:ssXXX</string>
       </void>
      </array>
     </void>
     <void property="directoryNameRelative">
      <string>/tulinkry.git.5</string>
     </void>
     <void property="parent">
      <string>https://github.com/tulinkry/git</string>
     </void>
     <void property="type">
      <string>git</string>
     </void>
    </object>
   </void>
   <void method="add">
    <object class="org.opengrok.indexer.history.RepositoryInfo">
     <void property="branch">
      <string>master</string>
     </void>
     <void property="currentVersion">
      <string>2018-10-22 19:49 +0200 675395c Kryštof Tulinger fixing composer.json</string>
     </void>
     <void property="datePatterns">
      <array class="java.lang.String" length="1">
       <void index="0">
        <string>yyyy-MM-dd&apos;T&apos;HH:mm:ssXXX</string>
       </void>
      </array>
     </void>
     <void property="directoryNameRelative">
      <string>/tulinkry.git.4</string>
     </void>
     <void property="parent">
      <string>https://github.com/tulinkry/git</string>
     </void>
     <void property="type">
      <string>git</string>
     </void>
    </object>
   </void>
  </void>
  <void property="sourceRoot">
   <string>/private/tmp/opengrok/src</string>
  </void>
 </object>
</java>
```

and shows that the state is as expected (no doubled repositories). When I restart tomcat, the problem vanish. My thoughts lead to `project_repository_map` and locking around it.

### PR DESCRIPTION
**Pull request by tulinkry**
fixes #2445

I run the test on the setup as for #2423, for 400 times in a row and didn't encounter any incosistency in terms of repositories on the main page.

### CODE REVIEW

**Comment by vladak** [comment]
The `configLock` is meant to lock the `configuration` only. Rather than using the lock for another piece of data, introduce separate lock for the `RuntimeEnvironment` members.

**Comment by tulinkry** [comment]
I reworked the code. I even added a test case. This seems to be enough locking

File: opengrok-indexer/src/main/java/org/opengrok/indexer/configuration/RuntimeEnvironment.java [hunk 1]
```diff
@@ -1432,21 +1431,25 @@ public void signalTorefreshSearcherManagers(List<String> subFiles, String host)
...
-            Project proj;
-            String repoPath;
-            try {
-                repoPath = PathUtils.getPathRelativeToSourceRoot(
-                    new File(r.getDirectoryName()), 0);
-            } catch (ForbiddenSymlinkException e) {
-                LOGGER.log(Level.FINER, e.getMessage());
-                continue;
-            }
+        synchronized (repository_map) {
```
Comment (vladak) [inline opengrok-indexer/src/main/java/org/opengrok/indexer/configuration/RuntimeEnvironment.java]: oh yes

File: opengrok-web/src/test/java/org/opengrok/web/api/v1/controller/ConfigurationControllerTest.java [hunk 2]
```diff
@@ -291,4 +297,70 @@ public void run() {
...
+            project.setPath("/project-" + i);
+            RepositoryInfo repo = new RepositoryInfo();
+            repo.setDirectoryNameRelative("/project-" + i);
+
+            projects.put("project-" + i, project);
+            repositoryInfos.add(repo);
+
+            // create the repository
+            FileUtils.copyDirectory(
+                    Paths.get(repository.getSourceRoot(), "git").toFile(),
```
Comment (vladak) [inline opengrok-web/src/test/java/org/opengrok/web/api/v1/controller/ConfigurationControllerTest.java]: I think this necessitates the annotation of the test case with git.

File: opengrok-web/src/test/java/org/opengrok/web/api/v1/controller/ConfigurationControllerTest.java [hunk 3]
```diff
@@ -291,4 +297,70 @@ public void run() {
         // Revert the value back to the default.
         env.setHitsPerPage(origValue);
     }
+
+    @Test
+    public void test() throws InterruptedException, IOException {
+        final int nThreads = 40;
```
Comment (vladak) [inline opengrok-web/src/test/java/org/opengrok/web/api/v1/controller/ConfigurationControllerTest.java]: possibly make the 2 constants based on number of CPUs in the system ?

Comment (tulinkry) [reply opengrok-web/src/test/java/org/opengrok/web/api/v1/controller/ConfigurationControllerTest.java]: I added the second constant and made the nThreads dependent on the testing system cpu

**Pull request merged**